cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -304,46 +361,77 @@ public boolean apply(RunRecord input) {
       .add(program.getType().name())
       .add(program.getId())
       .build();
+
     return list(activeKey, null, RunRecord.class, limit,
                 new Predicate<RunRecord>() {
                   @Override
                   public boolean apply(RunRecord input) {
-                    return input.getStartTs() >= startTime && input.getStartTs() < endTime;
+                    boolean normalCheck = input.getStartTs() >= startTime && input.getStartTs() < endTime;
+                    // If RunRecord has a valid adapterName, then it should match the adapter name passed in.
+                    if (normalCheck && adapter != null) {
+                      normalCheck = adapter.equals(input.getAdapterName());
+                    }
+                    return normalCheck;
                   }
                 });
   }
 
   private List<RunRecord> getHistoricalRuns(Id.Program program, ProgramRunStatus status,
-                                            final long startTime, final long endTime, int limit) {
+                                            final long startTime, final long endTime, int limit, final String adapter) {
     MDSKey historyKey = new MDSKey.Builder().add(TYPE_RUN_RECORD_COMPLETED,
                                                  program.getNamespaceId(),
                                                  program.getApplicationId(),
                                                  program.getType().name(),
                                                  program.getId()).build();
+
     MDSKey start = new MDSKey.Builder(historyKey).add(getInvertedTsScanKeyPart(endTime)).build();
     MDSKey stop = new MDSKey.Builder(historyKey).add(getInvertedTsScanKeyPart(startTime)).build();
     if (status.equals(ProgramRunStatus.ALL)) {
       //return all records (successful and failed)
-      return list(start, stop, RunRecord.class, limit, Predicates.<RunRecord>alwaysTrue());
+      return list(start, stop, RunRecord.class, limit, new Predicate<RunRecord>() {
+        @Override
+        public boolean apply(@Nullable RunRecord record) {
+          boolean normalCheck = true;
```

Comment:
nit: don't need this variable, can just return true at the bottom and return the condition inside the if.